cpp/src/strings/split/split_re.cu

```diff
@@ -0,0 +1,374 @@
+                                rmm::cuda_stream_view stream,
+                                rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(!pattern.empty(), "Parameter pattern must not be empty");
+
+  auto const max_tokens    = maxsplit > 0 ? maxsplit : std::numeric_limits<size_type>::max();
+  auto const strings_count = input.size();
+
+  std::vector<std::unique_ptr<column>> results;
+  if (strings_count == 0) {
+    results.push_back(make_empty_column(type_id::STRING));
+    return std::make_unique<table>(std::move(results));
+  }
+
+  // create the regex device prog from the given pattern
+  auto d_prog = reprog_device::create(pattern, get_character_flags_table(), strings_count, stream);
+  auto d_strings = column_device_view::create(input.parent(), stream);
+
+  // count the number of delimiters matched in each string
+  auto offsets = count_matches(*d_strings, *d_prog, stream, rmm::mr::get_current_device_resource());
+  auto offsets_view = offsets->mutable_view();
+  auto d_offsets    = offsets_view.data<offset_type>();
+
+  // get the split tokens from the input column; this also converts the counts into offsets
+  auto tokens = generate_tokens(*d_strings, *d_prog, direction, max_tokens, offsets_view, stream);
+
+  // the output column count is the maximum number of tokens generated for any input string
+  auto const columns_count = thrust::transform_reduce(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(strings_count),
+    [d_offsets] __device__(auto const idx) -> size_type {
+      return d_offsets[idx + 1] - d_offsets[idx];
+    },
+    0,
+    thrust::maximum<size_type>{});
+
+  // boundary case: if no columns, return one all-null column (custrings issue #119)
+  if (columns_count == 0) {
+    results.push_back(std::make_unique<column>(
+      data_type{type_id::STRING},
+      strings_count,
+      rmm::device_buffer{0, stream, mr},  // no data
+      cudf::detail::create_null_mask(strings_count, mask_state::ALL_NULL, stream, mr),
+      strings_count));
+    return std::make_unique<table>(std::move(results));
+  }
+
+  // convert the tokens into multiple strings columns
+  auto make_strings_lambda = [&](size_type column_index) {
+    // returns appropriate token for each row/column
+    auto indices_itr = cudf::detail::make_counting_transform_iterator(
+      0, tokens_transform_fn{*d_strings, tokens.data(), d_offsets, column_index});
+    return make_strings_column(indices_itr, indices_itr + strings_count, stream, mr);
+  };
+  // build a vector of columns
+  results.resize(columns_count);
+  std::transform(thrust::make_counting_iterator<size_type>(0),
+                 thrust::make_counting_iterator<size_type>(columns_count),
+                 results.begin(),
+                 make_strings_lambda);
+
+  return std::make_unique<table>(std::move(results));
+}
+
+std::unique_ptr<column> split_record_re(strings_column_view const& input,
```

Comment:
This function's implementation is so nice to read, because all the complicated bits are abstracted into helper functions and everything has clear names and comments. Just wanted to say this made me happy to review!